commit 8f2803715e92def1c03316b25680901e12fbeb9b
Author: Bernhard Manfred Gruber <bernhardmgruber@gmail.com>
Date:   Thu Jul 9 22:58:33 2026 +0200

    Do not vectorize large type reductions (#9762)
    
    Fixes: #9761

diff --git a/cub/cub/agent/agent_reduce.cuh b/cub/cub/agent/agent_reduce.cuh
index 24b58a8dea..b78a26d723 100644
--- a/cub/cub/agent/agent_reduce.cuh
+++ b/cub/cub/agent/agent_reduce.cuh
@@ -222,8 +222,12 @@ struct AgentReduceImpl
   static constexpr bool ATTEMPT_VECTORIZATION =
     (vec_size > 1) && (ITEMS_PER_THREAD % vec_size == 0)
     && (::cuda::std::is_pointer_v<InputIteratorT>)
-    // TODO(bgruber): remove the check for is_primitive<ValueT> in CCCL 4.0
-    &&(is_primitive<InputT>::value || THRUST_NS_QUALIFIER::is_trivially_relocatable_v<InputT>);
+         // TODO(bgruber): remove the check for is_primitive<ValueT> in CCCL 4.0
+         &&(is_primitive<InputT>::value || THRUST_NS_QUALIFIER::is_trivially_relocatable_v<InputT>)
+         // vectorizing large types leads to regressions again, see https://github.com/NVIDIA/cccl/issues/9761
+         // TODO(bgruber): this should be decided by tuning
+         &&sizeof(InputT)
+         <= 8;
 
   static constexpr CacheLoadModifier LOAD_MODIFIER = AgentReducePolicy::LOAD_MODIFIER;
 
