--- a/libcudacxx/include/cuda/std/__algorithm/lower_bound.h
+++ b/libcudacxx/include/cuda/std/__algorithm/lower_bound.h
@@ -30,17 +30,21 @@
 #include <cuda/std/__iterator/iterator_traits.h>
 #include <cuda/std/__type_traits/is_callable.h>
 #include <cuda/std/__type_traits/remove_reference.h>
+#include <cuda/std/__type_traits/is_integral.h>
+#include <cuda/std/__type_traits/is_signed.h>
+#include <cuda/std/__type_traits/is_same.h>
+#include <cuda/std/cstdint>
+#include <cuda/std/limits>
 
 #include <cuda/std/__cccl/prologue.h>
 
 _CCCL_BEGIN_NAMESPACE_CUDA_STD
 
 _CCCL_EXEC_CHECK_DISABLE
-template <class _AlgPolicy, class _Iter, class _Sent, class _Type, class _Proj, class _Comp>
-_CCCL_API constexpr _Iter __lower_bound(_Iter __first, _Sent __last, const _Type& __value, _Comp& __comp, _Proj& __proj)
+template <class _AlgPolicy, class _Iter, class _Distance, class _Type, class _Proj, class _Comp>
+_CCCL_API constexpr _Iter __lower_bound_counted_local(
+  _Iter __first, _Distance __len, const _Type& __value, _Comp& __comp, _Proj& __proj)
 {
-  auto __len = _IterOps<_AlgPolicy>::distance(__first, __last);
-
   while (__len != 0)
   {
     auto __l2 = ::cuda::std::__half_positive(__len);
@@ -57,6 +61,26 @@
     }
   }
   return __first;
+}
+_CCCL_EXEC_CHECK_DISABLE
+template <class _AlgPolicy, class _Iter, class _Sent, class _Type, class _Proj, class _Comp>
+_CCCL_API constexpr _Iter __lower_bound(_Iter __first, _Sent __last, const _Type& __value, _Comp& __comp, _Proj& __proj)
+{
+  auto __len = _IterOps<_AlgPolicy>::distance(__first, __last);
+  using _Distance = decltype(__len);
+  if constexpr (is_same_v<_AlgPolicy, _ClassicAlgPolicy>)
+  {
+    if constexpr (__has_random_access_traversal<_Iter> && is_integral_v<_Distance> && is_signed_v<_Distance>
+                  && numeric_limits<_Distance>::digits > 31)
+    {
+      if (__len >= 0 && __len <= static_cast<_Distance>(2147483647))
+      {
+        return ::cuda::std::__lower_bound_counted_local<_AlgPolicy>(
+          __first, static_cast<int32_t>(__len), __value, __comp, __proj);
+      }
+    }
+  }
+  return ::cuda::std::__lower_bound_counted_local<_AlgPolicy>(__first, __len, __value, __comp, __proj);
 }
 
 template <class _ForwardIterator, class _Tp, class _Compare>
